Emoji panel and clipboard history: locate actual items when traversing the elements for element selected event

josephsl opened the pull request:
Hi,

A reboot of #10378 with updated assumptions and lessons learned:

### Link to issue number:
Fixes #10377

### Summary of the issue:
In Version 1903 and later, on some systems, when emoji panel and clipboard history opens, wrong items are announced.

### Description of how this pull request fixes the issue:
Refined element traversal strategies when emoji panel and clipboard history opens. Also performed lints and changed cachedAutomationID to UIAAutomationId, along with using dedicated input panel and input panel Automation Id variables to react to window open event.

### Testing performed:
Tested via Windows 10 App Essentials for several months:

1. Opening emoji panel multiple times.
2. Opening clipboard history.
3. Toggling between emoji panel and clipboard history without closing them.
4. Searching for emoji items.
5. Reopening emoji panel after selecting People emoji group in Version 1903 and later.

### Known issues with pull request:
There is no way of telling NVDA that modern input elements are shown on screen (that's reserved for a future pull request; see #11454 for details).

### Change log entry:
Same as #10378.

Thanks.

josephsl committed:
Modern keyboard: update copyright header (year and removed intro text for consistency with other files)

josephsl committed:
Modern keyboard/window open event: use a dedicated input panel object and input panel Automation Id to react to various parts of modern input facility.

A dedicated input panel element will be used as a starting point for traversing modern keyboard interface. Along with this, rename childAutomationID to input panel Automation Id to better communicate the fact that window open event will be using Automation Id to detect different modern input features and react accordingly.

josephsl committed:
Modern keyboard: refine element traversal method for emoji panel and clipboard history to account for changes in Version 1903 on some systems. Re #10377.

Refined element traversal in window open event:
* Emoji panel: in Version 1903 (and on some systems), something other than emoji items are selected, particularly when People emoji group opens and skin tone modifier is somehow selected.
* Clipboard history: if clipboard is empty, clipboard status text isn't announced due to moving to a different element.

josephsl committed:
Modern keyboard/emoji panel and clipboard history: locate items themselves rather than parent lists. Re #10377.

Rather than calling element selected event on emojis list or clipboard history list, descend one more level (or in case of People emoji group, actual emojis list) so the event can deal with items themselves. This also means element selected event won't have to worry about traversing the elements tree one more time.

josephsl committed:
Modern keyboard: convert element selected event call to eventHandler.executeEvent.

Rather than executing UIA element selected event directly, use event handler's execute event function to schedule this event. This allows event handling features from NVDA to respond to this event.

josephsl committed:
Modern keyboard: separate classic emoji panel Automation Id's into a separate list, along with using isWin10 function.

Rather than using a build number, use winVersion.isWin10 function to check Windows 10 releases. Also, separate 1709 and 1803 emoji panel Automation Id's to a separate list.

josephsl committed:
Modern keyboard: Flake8 lint

josephsl commented:
Blocked by #11562 due to edits to window open event handler.

Thanks.

leonardder commented:
Does this also fix the emoji panel on 21H1 builds?

josephsl commented:
Hi, the new design? Can’t verify that, as I didn’t get the new design yet (it is rolling out to a subset of dev Insiders). Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, October 7, 2020 10:17 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Emoji panel and clipboard history: locate actual items when traversing the elements for element selected event (#11485)

 

Does this also fix the emoji panel on 21H1 builds?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11485#issuecomment-705077865> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEANXWUWJHWKWIBZHUTSJSPAPANCNFSM4P3XIFPQ> .

josephsl committed:
Merged master

josephsl committed:
Modern keyboard: remove duplicate event handler import. Re #10377

josephsl committed:
Modern keyboard: rename inputPanel to firstChild for consistency with UIA IME candidate panel check. Re #10377.

IME candidate panel check already assigns a dedicated variable to obj.firstChild, making input panel variable assignment redundant. Also, rename input panel Automation Id back to childAutomationId.

josephsl commented:
Hi,
 Ready for review again now that #11562 is here.

Thanks.

feerrenrut commented:
I'm asking @leonardder for a review because he was also reviewing #10378

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -123,13 +122,8 @@ def event_UIA_elementSelected(self, obj, nextHandler):
+		# In build 18262, emoji panel may open to People group and skin tone modifier gets selected.
+		if obj.parent.UIAAutomationId == "SkinTonePanelModifier_ListView":
 			# But this will point to nothing if emoji search results are not people.
 			if obj.parent.next is not None: obj = obj.parent.next
```
Could you split this and the line below into multiple lines?

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -169,35 +169,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+			and config.conf["inputComposition"]["autoReportAllCandidates"]
+		):
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.firstChild.firstChild, nextHandler)
```
I rather see checking every child's boolean value rather than doing a try/except here. That's just my personal preference.

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -169,35 +169,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+		# Emoji panel in Version 1809 (specifically, build 17666) and later.
+		elif childAutomationId == "TEMPLATE_PART_ExpressionGroupedFullView":
 			self._emojiPanelJustOpened = True
+			# #10377: on some systems, there is something else besides grouping controls,
```
Could you elaborate on what you mean with 'some systems'?

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -169,35 +169,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
 			try:
-				self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+				# Avoid announcing skin tone modifiers if possible.
+				emojiItem = emojisList.firstChild.firstChild
```
Again, are we sure that there is a first child that has a first child? I'm afraid this can lead to breakage when this changes in future.

josephsl replied:
Yes, for now emoji panel consists of the emoji list, a grouping (first child), and the actual emojis list unless people emoji is selected, in which case the third level will be skin tone modifiers list object. We need to come back to this if we get a confirmation that the new modern keyboard UI significantly changes this. The upcoming commit will check if the first child object is present for each level as suggested.

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -169,35 +169,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+				# Avoid announcing skin tone modifiers if possible.
+				emojiItem = emojisList.firstChild.firstChild
+				if emojiItem.UIAAutomationId == "SkinTonePanelModifier_ListView":
+					emojiItem = emojiItem.next
```
Same as above

leonardder commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -169,35 +169,55 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+		elif childAutomationId == "TEMPLATE_PART_ClipboardTitleBar":
+			# Under some cases, clipboard tip text isn't shown on screen,
+			# causing clipboard history title to be announced instead of most recently copied item.
+			clipboardHistoryItem = obj.children[-2]
```
Same as above

LeonarddeR reviewed (changes requested):
Could you please especially consider cases where you're traversing objects without checking whether the objects exist?

josephsl commented:
Hi, try/except was done for compactness and for readability. I admit that the code under review came from my add-on, so I’ll do something about these from both sides. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Tuesday, October 13, 2020 10:09 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Emoji panel and clipboard history: locate actual items when traversing the elements for element selected event (#11485)

 

@leonardder requested changes on this pull request.

Could you please especially consider cases where you're traversing objects without checking whether the objects exist?

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/11485#discussion_r504116602> :

> @@ -123,13 +122,8 @@ def event_UIA_elementSelected(self, obj, nextHandler):
               if obj.UIAElement.cachedAutomationID == "CandidateList": return
               speech.cancelSpeech()
               # Sometimes, due to bad tree traversal or wrong item getting selected, something other than the selected item sees this event.
-              # Sometimes clipboard candidates list gets selected, so ask NvDA to descend one more level.
-              if obj.UIAElement.cachedAutomationID == "TEMPLATE_PART_ClipboardItemsList":
-                      obj = obj.firstChild
-              # In build 18262, emoji panel may open to People group and skin tone modifier or the list housing them gets selected.
-              elif obj.UIAElement.cachedAutomationID == "SkinTonePanelModifier_ListView":
-                      obj = obj.next
-              elif obj.parent.UIAElement.cachedAutomationID == "SkinTonePanelModifier_ListView":
+              # In build 18262, emoji panel may open to People group and skin tone modifier gets selected.
+              if obj.parent.UIAAutomationId == "SkinTonePanelModifier_ListView":
                       # But this will point to nothing if emoji search results are not people.
                       if obj.parent.next is not None: obj = obj.parent.next

Could you split this and the line below into multiple lines?

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/11485#discussion_r504118094> :

>                      try:
-                              self.event_UIA_elementSelected(obj.firstChild.firstChild.firstChild, nextHandler)

I rather see checking every child's boolean value rather than doing a try/except here. That's just my personal preference.

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/11485#discussion_r504118458> :

>                      self._emojiPanelJustOpened = True
+                      # #10377: on some systems, there is something else besides grouping controls,

Could you elaborate on what you mean with 'some systems'?

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/11485#discussion_r504119559> :

>                      try:
-                              self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+                              # Avoid announcing skin tone modifiers if possible.
+                              emojiItem = emojisList.firstChild.firstChild

Again, are we sure that there is a first child that has a first child? I'm afraid this can lead to breakage when this changes in future.

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/11485#discussion_r504119700> :

>                      try:
-                              self.event_UIA_elementSelected(obj.firstChild.children[-2].firstChild.firstChild, nextHandler)
+                              # Avoid announcing skin tone modifiers if possible.
+                              emojiItem = emojisList.firstChild.firstChild
+                              if emojiItem.UIAAutomationId == "SkinTonePanelModifier_ListView":
+                                     emojiItem = emojiItem.next

Same as above

  _____  

In source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py <https://github.com/nvaccess/nvda/pull/11485#discussion_r504120018> :

>                      except AttributeError:
                               # In build 18272's emoji panel, emoji list becomes empty in some situations.
                               pass
               # Clipboard history.
               # Move to clipboard list so element selected event can pick it up.
               # #9103: if clipboard is empty, a status message is displayed instead, and luckily it is located where clipboard data items can be found.
-              elif childAutomationID == "TEMPLATE_PART_ClipboardTitleBar":
-                      self.event_UIA_elementSelected(obj.children[-2], nextHandler)
+              elif childAutomationId == "TEMPLATE_PART_ClipboardTitleBar":
+                      # Under some cases, clipboard tip text isn't shown on screen,
+                      # causing clipboard history title to be announced instead of most recently copied item.
+                      clipboardHistoryItem = obj.children[-2]

Same as above

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/11485#pullrequestreview-507652116> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXECXDBO6SCXZRYQ7WYLSKSCRRANCNFSM4P3XIFPQ> .

josephsl committed:
Modern keyboard: address review comments on multiple statements and comment clarification. Re #10377.

Reviewed by Leonard de Ruijter: split lines with multiple statements, clarify that 'some systems' refer to non-English builds of Version 1903 (May 2019 Update) and later.

josephsl committed:
Modern keyboard: check if emoji panel list/grouping/list items are present. Re #10377.

Suggested by Leonard de Ruijter: when assigning emoji item, check each emoji panel level (emoji list header, grouping, emoji items/skin tone modifier list) to make sure child objects are present (not None). In case skin tone modifiers list is selected, before moving to the next object (emoji item), make sure it is even present. This avoids a try/except statement and is used to force NVDA to announce something rather than staying silent of attribute error is raised in the future due to modern keyboard UI changes.

josephsl committed:
Modern keyboard: eventHandler.executeEvent -> queueEvent for consistency with UIA IME candidates window.

josephsl commented:
Hi,

It was pointed out a few days ago that the emoji panel was redesigned in a dev channel WIP build. I'm sorry to inform everyone that the redesigned emoji panel cannot be used with NVDA at all - no announcements with both Narrator and NVDA when the panel opens, so for the purposes of this PR, the redesigned emoji panel is out of scope.

Thanks.

leonardder commented:
@josephsl I'm on the most recent insider preview build, and the panel seems to work correct again. Does this pr needs to be revised for this? At least, I think I'd prefer it if you check the current changes against an insider build before continuing with this.

josephsl commented:
Hi, nope. There is an issue with dev channel build that has to do with categories firing element selected event but that can be handed by Windows 10 App Essentials add-on and is beyond the PR scope. Thanks.

josephsl commented:
Hi all,

Ready for another review if needed. Thanks.

josephsl committed:
Merge branch 'master' into i10377

josephsl commented:
Hi,

Update: build 21277 (rs_prerelease, released December 10) brought back modern emoji panel. As for my assessment on it, see the above comment (brought screen readers to their knees again). Therefore for the purposes of this PR, the new emoji panel will not be supported in this PR.

Thanks.

josephsl commented:
Hi,

March 2021 update: the new emoji panel (discussed above) is usable, although with some refinements required (it uses UIA notification event to "announce" selected emoji). Because the UI is different than what we have, it won't be part of this PR.

Thanks.

josephsl committed:
Merge branch 'master' into i10377

josephsl committed:
Merge branch 'master' into i10377

josephsl committed:
Merge branch 'master' into i10377

josephsl commented:
Hi,

One more change: to prepare for Python 3.10's pattern matching syntax and readability, when checking for 1709 and 1803 emoji panel, Automation Id will be checked before build range; no changes to semantics whatsoever.

Thanks.

josephsl committed:
Merge branch 'master' into i10377

josephsl committed:
Modern keyboard/emoji panel in 1709 and 1803: check Automation Id before build range for consistency with the rest of UIA window open event.

seanbudd commented:
closing/reopening to trigger a build

seanbudd commented:
Thanks @josephsl,

Would you be able to add system tests to smoke test the emoji panel? I understand your capacity is changing right now (congratulations!) so let us know if the work needs to be passed over.

josephsl commented:
Hi,

I might not be able to add system tests to smoke this. The biggest concern is the fact that emoji panel UI is changing in Cobalt releases (build 21300 and later) which will cause tests to fail unless winVersion.getWinVer().build is checked first.

The test procedure should be:

1. Check if emoji panel is present (winVersion.GetWinVer() >= winVersion.WIN10_1809; Version 1709 is out of support and emoji panel UI has changed in Version 1809, that's why). This will need to be refined if adding smoke tests for Cobalt.
2. Press Windows+Period to open emoji panel.
3. Use the object navigation path as specified in this PR and see if NVDA announces emoji of any kind (you can't really rely on the exact emoji being announced because the emoji panel may learn to give you something else; I expect it won't be a big concern as Appveyor spawns a fresh virtual machine every time a commit is made). A possible check is making sure the resulting navigator object is indeed a UIA object with a specific Automation Id and/or class name and/or is part of a specific app module (you can check for api.getNavigatorObject().appName == "textinputhost" provided that the VM runs Version 2004 (build 19041) or later; if it runs Server 2019 (Version 1809) or 190x (1903/1909), then it becomes a bit complicated as the executable name is rather long).

NOtes on Cobalt emoji panel:

1. Emoji panel and clipboard history have been combined under a web document interface.
2. UIA notification event, not element selected event, is used to announce emojis. Therefore object navigation strategy won't work.

Making matters even more complicated is a possibility that Version 21H2 client and server code might be different - currently latest Server 2022 preview (Version 21H2) code is based on work done last year (iron/fe_release/build 203xx), whereas client is based on work being done this year (cobalt/co_release/build 213xx). If it turns out Version 21H2 code between client and server are different, and if Server 2022 will be based on Iron (build 203xx), then it might be a bit easier to smoke test emoji panel (although this means dealing with an extra UIA element selected event being fired somehow); all this is thrown out the window if Server 2022 will be based on Cobalt and Appveyor decides to use Server 2022 VM, in which case smoke test cannot be added.

I know how complicated the picture is (this is one of the personal gripes with what's happening at Microsoft build labs, but such things are expected; this is one of those perks and stresses of being a Windows Insider and trying to guess what Microsoft will do in the short term).

Hope this helps. Thanks.

seanbudd commented:
Hi @josephsl,

Thanks for the detailed notes. I think for now we can just write a system that passes for Iron (assuming that it would currently work for AppVeyor and personal machines). If the system test starts to fail we can exclude it from builds so that AppVeyor continues to pass, and prioritise work to support Cobalt if it hasn't been done yet.

josephsl commented:
Hi, I expect Cobalt won’t be finalized until mid-June, so perhaps it would be best to defer support for Cobalt emoji panel until fall (2021.2) at the earliest, with system tests planned for summer (I think I might not be able to write tests, but would be able to help plan things a bit before starting graduate studies). Thanks.

seanbudd committed:
emoji panel and clipboard system tests

seanbudd committed:
fix teardown

seanbudd committed:
exclude panel search from appveyor build

seanbudd committed:
fix behaviour to be more expected

seanbudd committed:
exclude emoji test from build

josephsl commented:
Blocked by #12487. Thanks.

seanbudd committed:
more explicit tags for exclusion

seanbudd committed:
use run dialog instead of search bar

seanbudd committed:
Merge branch 'master' into i10377

seanbudd reviewed (approved):
This seems to be somewhat unnecessarily blocked by things that may take a while to resolve. It would be good to still merge the system tests in #12487, but I have tested this PR with the tests and it seems okay.

seanbudd committed:
update changes

josephsl commented:
Hi, agreed about system tests – let’s get the testing party started with alpha builds!!! Thanks (this will lessen the burden on Windows App Essentials add-on significantly).

seanbudd committed:
Merge branch 'master' into i10377
